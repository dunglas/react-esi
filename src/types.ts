export interface IWebpackProcess extends NodeJS.Process {
  browser?: boolean;
}

export type GlobalProcess = typeof globalThis & {
  process: IWebpackProcess;
  __REACT_ESI__: { [s: string]: object };
};
