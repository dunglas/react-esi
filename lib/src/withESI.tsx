import PropTypes from "prop-types";
import type { WeakValidationMap, ComponentType, ComponentClass } from "react";
import React from "react";

declare global {
  interface Window {
    __REACT_ESI__: { [s: string]: object };
  }

  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace NodeJS {
    interface Process {
      browser?: boolean;
    }
  }
}

interface IWithESIProps {
  esi?: {
    attrs?: {
      [key: string]: string | null;
    };
  };
}

// Prevent bundlers to bundle server.js
const safeRequireServer = () => {
  try {
    // Necessary for NextJS
    return eval("require('react-esi/lib/server')");
  } catch (error) {
    // Necessary for Express and others
    return eval("require('./server')");
  }
};

const isClient = () => {
  return (
    (typeof process !== "undefined" && process?.browser) ||
    typeof window !== "undefined"
  );
};

const isServer = () => !isClient();

interface State {
  childProps: object;
  initialChildPropsLoaded: boolean;
}
/**
 * Higher Order Component generating a <esi:include> tag server-side,
 * and rendering the wrapped component client-side.
 */
export default function withESI<P>(
  WrappedComponent: ComponentType<P>,
  fragmentID: string
): ComponentClass<IWithESIProps & P> {
  return class WithESI extends React.Component<P & IWithESIProps, State> {
    public static WrappedComponent = WrappedComponent;
    public static displayName = `WithESI(${
      WrappedComponent.displayName || WrappedComponent.name || "Component"
    })`;
    public static propTypes = {
      esi: PropTypes.shape({
        attrs: PropTypes.objectOf(PropTypes.string), // extra attributes to add to the <esi:include> tag
      }),
    } as unknown as WeakValidationMap<IWithESIProps & P>;
    public state: State = {
      childProps: {},
      initialChildPropsLoaded: true,
    };
    private esi = {};

    constructor(props: P & IWithESIProps) {
      super(props);
      const { esi, ...childProps } = props;
      this.esi = esi || {};
      this.state.childProps = childProps;

      if (isServer()) {
        return;
      }

      if (window.__REACT_ESI__?.[fragmentID]) {
        // Inject server-side computed initial props
        this.state.childProps = {
          ...window.__REACT_ESI__[fragmentID],
          ...this.state.childProps,
        };
        return;
      }

      // TODO: add support for getServerSideProps
      if ("getInitialProps" in WrappedComponent) {
        // No server-side rendering for this component, getInitialProps will be called during componentDidMount
        this.state.initialChildPropsLoaded = false;
      }
    }

    public componentDidMount() {
      if (this.state.initialChildPropsLoaded) {
        return;
      }

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (WrappedComponent as any)
        .getInitialProps({ props: this.state.childProps })
        .then((initialProps: object) =>
          this.setState({
            childProps: initialProps,
            initialChildPropsLoaded: true,
          })
        );
    }

    public render() {
      if (isClient()) {
        return (
          <WrappedComponent
            {...(this.state.childProps as JSX.IntrinsicAttributes & P)}
          />
        );
      }

      const server = safeRequireServer();

      return server.createIncludeElement(fragmentID, this.props, this.esi);
    }
  };
}
