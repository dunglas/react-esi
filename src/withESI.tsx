import PropTypes from "prop-types";
import React, { WeakValidationMap } from "react";

declare global {
  interface Window {
    __REACT_ESI__: { [s: string]: object };
  }
}

interface IWebpackProcess extends NodeJS.Process {
  browser?: boolean;
}

interface IWithESIProps {
  esi?: {
    attrs?: {
      [key: string]: string | null;
    };
  };
}

/**
 * Higher Order Component generating a <esi:include> tag server-side, and rendering the wrapped component client-side.
 */
export default function withESI<P>(
  WrappedComponent: React.ComponentType<P>,
  fragmentID: string
): React.ComponentClass<IWithESIProps & P> {
  return class WithESI extends React.Component<P & IWithESIProps> {
    public static WrappedComponent = WrappedComponent;
    public static displayName = `WithESI(${
      WrappedComponent.displayName || WrappedComponent.name || "Component"
    })`;
    public static propTypes = {
      esi: PropTypes.shape({
        attrs: PropTypes.objectOf(PropTypes.string) // extra attributes to add to the <esi:include> tag
      })
    } as unknown as WeakValidationMap<IWithESIProps & P>;
    public state = {
      childProps: {},
      initialChildPropsLoaded: true
    };
    private esi = {};

    constructor(props: P & IWithESIProps) {
      super(props);
      const { esi, ...childProps } = props;
      this.esi = esi || {};
      this.state.childProps = childProps;

      if (!(process as IWebpackProcess).browser) {
        return;
      }

      if (window.__REACT_ESI__ && window.__REACT_ESI__[fragmentID]) {
        // Inject server-side computed initial props
        this.state.childProps = {
          ...window.__REACT_ESI__[fragmentID],
          ...this.state.childProps
        };
        return;
      }

      // TODO: add support for getServerSideProps
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      if ((WrappedComponent as any).getInitialProps) {
        // No server-side rendering for this component, getInitialProps will be called during componentDidMount
        this.state.initialChildPropsLoaded = false;
      }
    }

    public componentDidMount() {
      if (this.state.initialChildPropsLoaded) {
        return;
      }

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (WrappedComponent as any)
        .getInitialProps({ props: this.state.childProps })
        .then((initialProps: object) =>
          this.setState({
            childProps: initialProps,
            initialChildPropsLoaded: true
          })
        );
    }

    public render() {
      if ((process as IWebpackProcess).browser) {
        return (
          <div>
            <WrappedComponent {...(this.state.childProps as React.JSX.IntrinsicAttributes&P)} />
          </div>
        );
      }

      // Prevent Webpack and other bundlers to ship server.js
      const server = eval('require("./server")');
      return (
        <div
          dangerouslySetInnerHTML={{
            __html: server.createIncludeElement(
              fragmentID,
              this.props,
              this.esi
            )
          }}
        />
      );
    }
  };
}
