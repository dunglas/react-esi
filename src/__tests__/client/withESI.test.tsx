import React from "react";
import renderer from "react-test-renderer";
import withESI from "../../withESI";
import { GlobalProcess } from "../../types";


declare let global: GlobalProcess

const Dummy = (props: { name?: string }) => <div>Hello {props.name}</div>;

test("exposes WrappedComponent", () => {
  const DummyESI = withESI(Dummy, "id");
  expect(DummyESI).toHaveProperty("WrappedComponent", Dummy);
});

test("client-side", () => {
  const DummyESI = withESI(Dummy, "id");
  expect(DummyESI.displayName).toBe("WithESI(Dummy)");

  global.process.browser = true;
  const component = renderer.create(<DummyESI name="Kévin" />);
  expect(component).toMatchSnapshot();
});

test("client-side with serialized props", () => {
  const DummyESI = withESI(Dummy, "id");
  expect(DummyESI.displayName).toBe("WithESI(Dummy)");

  global.process.browser = true;
  global.__REACT_ESI__  = { id: { name: "Anne" } };
  const component = renderer.create(<DummyESI />);
  expect(component).toMatchSnapshot();
});

test("client-side call getInitialProps", async () => {
  let called = false;

  const Component = (props: { name?: string }) => <div>Hello {props.name}</div>;
  Component.getInitialProps = async () => {
    called = true;
    return { name: "Kévin" };
  };

  const ComponentESI = withESI(Component, "initial-props");

  global.process.browser = true;
  renderer.create(<ComponentESI />);
  expect(called).toBe(true);
});