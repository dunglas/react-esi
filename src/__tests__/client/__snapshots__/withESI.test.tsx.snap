// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client-side 1`] = `
<div>
  Hello 
  KÃ©vin
</div>
`;

exports[`client-side with serialized props 1`] = `
<div>
  Hello 
  Anne
</div>
`;
